vcd/resource_generic_crud.go

```diff
@@ -0,0 +1,192 @@
+package vcd
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/vmware/go-vcloud-director/v3/govcd"
+	"github.com/vmware/go-vcloud-director/v3/util"
+)
+
+// crudConfig defines a generic approach for managing Terraform resources where the parent entity is
+// a standard OpenAPI entity and the outer entity should satisfy 'updateDeleter' type constraint
+// (have 'Update' and 'Delete' pointer receiver methods)
+type crudConfig[O updateDeleter[O, I], I any] struct {
+	// entityLabel to use
+	entityLabel string
+
+	// getTypeFunc is responsible for converting schema fields to inner type
+	getTypeFunc func(d *schema.ResourceData) (*I, error)
+	// stateStoreFunc is responsible for storing state
+	stateStoreFunc func(d *schema.ResourceData, outerType O) error
+
+	// createFunc is the function that can create an outer entity based on inner entity config
+	// (which is created by 'getTypeFunc')
+	createFunc func(config *I) (O, error)
+
+	// resourceReadFunc that will be executed from Create and Update functions
+	resourceReadFunc func(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics
+
+	// getEntityFunc is a function that retrieves the entity
+	// It will use ID for resources and Name for data sources
+	getEntityFunc func(idOrName string) (O, error)
+
+	// preCreateHooks will be executed before the entity is created
+	preCreateHooks []schemaHook
+
+	// preDeleteHooks will be executed before the entity is deleted
+	preDeleteHooks []outerEntityHook[O]
+
+	// readHooks that will be executed after the entity is read, but before it is stored in state
+	readHooks []outerEntityHook[O]
+}
+
+// updateDeleter is a type constraint to match only entities that have Update and Delete methods
+type updateDeleter[O any, I any] interface {
+	Update(*I) (O, error)
+	Delete() error
+}
+
+// outerEntityHook defines a type for hook that can be fed into generic CRUD operations
+type outerEntityHook[O any] func(O) error
+
+// schemaHook defines a type for hook that can be fed into generic CRUD operations
+type schemaHook func(*VCDClient, *schema.ResourceData) error
+
+func createResource[O updateDeleter[O, I], I any](ctx context.Context, d *schema.ResourceData, meta interface{}, c crudConfig[O, I]) diag.Diagnostics {
+	t, err := c.getTypeFunc(d)
+	if err != nil {
+		return diag.Errorf("error getting %s type: %s", c.entityLabel, err)
```

Comment:
Should we state that it happened during Create? Just to differentiate between Create and Update.
```suggestion
		return diag.Errorf("error getting %s type on create: %s", c.entityLabel, err)
```

Reply:
Changed.